### PR TITLE
EOSIO onchain validation tool

### PR DESCRIPTION
**Pull request by winlin**

### CODE REVIEW

File: eos_validator.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            # Validate Failed because the eos_total NOT match all legal amount
+            raise Exception("There are illegal SYS transfer to unknow accounts")
+
+
+# Check the CONTRACT
+def check_contract(node_host="http://localhost:8888")
+    contract_hash_map = {
+        'eosio.msig':{'cmd':'cleos get code eosio.msig' 'abi':'YOUR_ABI_HASH_VAL', 'wasm':'YOUR_WASM_HASH_VAL'}
+        'eosio.token':{'cmd':'cleos get code eosio.token', 'abi':'YOUR_ABI_HASH_VAL', 'wasm':'YOUR_WASM_HASH_VAL'}
+        'eosio.disco':{'cmd':'cleos get code eosio.disco', 'abi':'YOUR_ABI_HASH_VAL', 'wasm':'YOUR_WASM_HASH_VAL'}
```
Comment (dfguo) [inline eos_validator.py:40]: For people who are not using `eos-bios` to sync the genesis and peer discovery process, this contract might not be there. So perhaps only enable it when a env var is set?

File: eos_validator.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        'eosio.unregd':{'cmd':'cleos get code eosio.unregd', 'abi':'YOUR_ABI_HASH_VAL', 'wasm':'YOUR_WASM_HASH_VAL'}
+        'eosio':{'cmd':'cleos get code eosio', 'abi':'YOUR_ABI_HASH_VAL', 'wasm':'YOUR_WASM_HASH_VAL'}
+        #... maybe more other contract
+    }
+    for contract in contract_hash_map:
```
Comment (dfguo) [inline eos_validator.py:45]: what about contracts that are not in hash map but secretly uploaded by someone

File: eos_validator.py (hunk 1 above)
Comment (winlin) [reply eos_validator.py:40]: yes, I will pass paramaters by a json file which you can modify to what you whant to check. The code is just for demonstration.

File: eos_validator.py (hunk 2 above)
Comment (winlin) [reply eos_validator.py:45]: that maybe happen, but I have NOT found method/api to get all the contract deployed onchain. So what I want to make sure is that the `eosio` account should NOT deploy other contract. And if others deploy contracts, I think the token holders will not have nothing to lose.

File: eos_validator.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
+
+# Check the TOKEN amount
+def check_balance(node_host="http://localhost:8888", snapshot_csv="./snapshot.csv", eos_total=1000000000):
+    with open(snapshot_csv, 'r') as fp:
+        account_onchain_total = 0
+        for onepersion in fp.readline()
```
Comment (sudoaza) [inline eos_validator.py:10]: This is ok for checking that all the buying got their EOS, but is missing checking that no other account was created/staked.

File: eos_validator.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            account_name, account_amount, pub_key = json.loads(onepersion)['name'], json.loads(onepersion)['amount'],json.loads(onepersion)['pub'] 
+            if account_amount<1.0:
+                continue
+            account_balance = exec('cleos -u ' + node_host + ' get currency balance eosio.token ' + account_name)
+            account_net_delegated = exec('cleos -u ' + node_host + ' get account ' + account_name)['delegated_bandwidth']['net_weight']
+            account_cpu_delegated = exec('cleos -u ' + node_host + ' get account ' + account_name)['delegated_bandwidth']['cpu_weight']
```
Comment (sudoaza) [inline eos_validator.py:16]: Here calls twice cleos get account on the same account, better save the output in a variable and then get the net_weight/cpu_weight from the variable, cut cleos calls by half

File: eos_validator.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            account_onchain_amount = account_balance+account_net_delegated+account_net_delegated
+            if abs(account_amount - account_onchain_amount) > 0.0001:
+                # Validate Failed because the amount in onchainblock NOT same with the snapshot
+                raise Exception("account balance failed: ", account_name)
+
+            account_onchain_pubkey = exec('cleos -u ' + node_host + ' get account ' + account_name)['permissions']['owner']
```
Comment (sudoaza) [inline eos_validator.py:22]: again? or am i missing somethign?

File: eos_validator.py (hunk 3 above)
Comment (winlin) [reply eos_validator.py:10]: yes, there is no method/api which can get all created accounts in the chain, so I just make sure the account from snapshot is correct. And the check of 
`if abs(eos_total - (account_onchain_total + eosio_onchain_balance)) > 0.0001:`
will make sure even there are extra account, but there will no token transfer to them.

File: eos_validator.py (hunk 4 above)
Comment (winlin) [reply eos_validator.py:16]: yes, the code is just for demo, I will do what you suggested when realize.

File: eos_validator.py (hunk 5 above)
Comment (winlin) [reply eos_validator.py:22]: compare the public key from snapshot with the acount onchain's public key to make sure transfer into correct erc20 mapping address

File: eos_validator.py (hunk 3 above)
Comment (sudoaza) [reply eos_validator.py:10]: ah cool

File: eos_validator.py (hunk 2 above)
Comment (dfguo) [reply eos_validator.py:45]: fair point

**Comment by winlin** [comment]
test pass

**Pull request merged**